---
title: "<span style='font-size: 16px; color: grey'> crisprQTL proof-of-concept experiment in T cells </span> <br> <span style='font-size: 36px'> Perturbation analysis </span>"
author: "Ximena Ibarra-Soria"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  html_document:
    theme: paper
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    df_print: paged
    code_folding: hide
    toc: true
    toc_depth: 4
    toc_float: 
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(scater)
library(scran)
library(dplyr)
library(scales)
library(BiocParallel)
library(BiocSingular)
library(ComplexHeatmap)
library(ggplot2)
library(ggpubr)
library(RColorBrewer)
library(inlmisc)
library(UpSetR)

dir <- paste0(getwd(), "/")

th <- theme_bw() + theme(
  axis.text.x = element_text(size=10), 
  axis.title.x = element_text(size=12), 
  axis.text.y = element_text(size=12), 
  axis.title.y = element_text(size=12),
  panel.grid.major = element_blank(), 
  panel.grid.minor = element_blank(), 
  axis.line = element_line(colour = "black"), 
  panel.border = element_blank(), 
  plot.title = element_text(face="bold", hjust = 0.5))

## configure parallelisation options
bpp <- MulticoreParam(38)

## FUNCTIONS
revlog_trans <- function(base = exp(10)){
  ## Define the desired transformation.
  trans <- function(x) {
    -log(x, base)
  }
  ## Define the reverse of the desired transformation
  inv <- function(x){
    base^(-x)
  }
  ## Creates the transformation
  trans_new(paste("revlog-", base, sep = ""),
            trans, ## The transformation function (can be defined using anonymous functions)
            inv,  ## The reverse of the transformation
            log_breaks(base = base), ## default way to define the scale breaks
            domain = c(1e-100, Inf) ## The domain over which the transformation is valued
  )
}
```

### Analysis strategy

Using the normalised data for good-quality cells with at least one gRNA assigned, we test the effects of each perturbation using several differential expression approaches:

- [`limma-voom`](https://doi.org/10.1186/gb-2014-15-2-r29): well-established methodology and extended to work with single-cell data.
- [`MAST`](https://doi.org/10.1186/s13059-015-0844-5): specifically developed for single-cell data.
- [`SCEPTRE`](https://doi.org/10.1186/s13059-015-0844-5): specifically developed for *high MOI* crisprQTL data.
- `wilcoxon rank sum test`: non-parametric test.

```{r data}
sce <- readRDS(paste0(dir, "results/03_sce_POC_Tcells.goodQual.calls.NORM.Rds"))
c(n_cells = ncol(sce))

# genes detected in at least 5% of the cells
genes_to_keep <- rowSums(counts(sce)>0) > ncol(sce)*0.05
genes_detected <- rownames(sce)[genes_to_keep]

## gRNA annotation
gRNA_ann <- read.table(paste0(dir, "data/gRNA_library.tsv"), sep="\t", header = TRUE)
```

The majority of the perturbations included in the library correspond to positive controls, where we have an expected effect:

- 35 gene **transcription start sites** (TSS). For these perturbations, we expect silencing of the targeted genes. Denoted `TSS`.
- 3 **locus control regions** (LCR) of CD2. Again, silencing these loci should lead to downregulation of *CD2* expression. Denoted `LCR`.
- 28 **intergenic loci** that have been **linked** to target genes by Gasperini et al. We expect decreased expression, but the caveat is that Gasperini et al. performed their screen in K562 cells, so we do not expect to recover all effects. Denoted `ENH`.

We detect most of the genes with an expected effect, except for:

```{r pos_ctrls_detected}
## check if all expected genes were detected
not_expr <- setdiff(unique(gRNA_ann$expected_DE_gene), c(genes_detected, NA))
not_expr <- unique(gRNA_ann[gRNA_ann$expected_DE_gene %in% not_expr,c(3:5)])
not_expr[order(not_expr$class, decreasing = TRUE),]
```

We additionally have 23 **intergenic loci** that overlap ENCODE cCREs. For these there is no hypothesis.

The library contains `r nrow(gRNA_ann)` gRNAs, each detected in a median of 282 cells. 

```{r cells_per_guide, fig.width=3, fig.height=3}
ncells_per_gRNA <- rowSums(counts(altExp(sce, 'gRNA_calls')))

ggplot(as.data.frame(ncells_per_gRNA), aes(1, ncells_per_gRNA)) +
  scale_y_log10() +
  geom_violin() +
  geom_boxplot(width=0.1) +
  annotation_logticks(side="l") +
  xlab("") +
  ylab("number of cells per gRNA") +
  th + theme(axis.text.x = element_blank(),
             axis.ticks.x = element_blank())

summary(ncells_per_gRNA)
```

If we aggregate all cells with any of the 4 gRNAs for each target, we have a median of 1,435 cells per target.

```{r cells_per_target, fig.width=3, fig.height=3}
df <- data.frame(gRNA = names(ncells_per_gRNA),
                 n_cells = ncells_per_gRNA,
                 target = gRNA_ann[match(names(ncells_per_gRNA), gRNA_ann$ID),]$target)
df <- df %>% group_by(target) %>%
  summarise(ncells_per_target = sum(n_cells))

ggplot(df, aes(1, ncells_per_target)) +
  scale_y_log10() +
  geom_violin() +
  geom_boxplot(width=0.1) +
  annotation_logticks(side="l") +
  xlab("") +
  ylab("number of cells per target") +
  th + theme(axis.text.x = element_blank(),
             axis.ticks.x = element_blank())

summary(df$ncells_per_target)
```

For the implementation of the algorithms:

- We focus the analysis on genes within 1Mb (up/downstream) of the target and that are detected in at least 5% of the cells. 
- Cells with a given gRNA are tested against a random sample of five thousand cells containing only non-targeting gRNAs, which represent the *wild-type* transcriptional state.
- For `limma` and `MAST` the model fit and dispersion estimates are performed using all genes (detected in at least 5% of cells), and all cells containing *any* of the four gRNAs from each target. The effect of each gRNA is then tested with the appropriate contrast.
- We include the number of genes detected per cell as a covariate.
- We block on the batch (the two independent experiments performed; see `03_normalisation_and_batch_correction.Rmd` for details).
- To assess p-value calibration, we also test sets of 4 non-targeting gRNAs against any gene within 1Mb of a target (i.e. 4 NT gRNAs create a mock new target for which we expect no differentially expressed genes). 


### Differential expression testing

For details on the implementation of each method, please see:

- `sceptre`: using the original release of the package (`v0.0.1`).
  + `04.1_gene_gRNA_pairs.R`: defining the genes in the vicinity of each target to be tested.
  + `04.2.1_sceptre_data_prep.R`: reformats data in the object types required.
  + `04.2.2_param_file_sceptre.R`: parameters used to run `sceptre`.
- `MAST`: 
  + `04.3_run_MAST.R`: run in parallel for subsets of targets.
- `limma-voom`:
  + `04.4_run_limma.R`
- `wilcoxon rank sum test`: non-parametric test.
  + `04.5_run_wilcoxong.R`: using the implementation from `scran`.

### Results

For all algorithms, we read in the results for all targets, and adjust the p-values for multiple testing (Benjamini & Hochberg) if not done already.

```{r sceptre}
sceptre <- readRDS(paste0(dir, "results/04_sceptre/results/all_results.rds"))
## adjust for multiple testing (per perturbation)
sceptre$FDR <- NA
for(guide in unique(sceptre$gRNA_id)){
  sceptre[sceptre$gRNA_id==guide,]$FDR <- p.adjust(sceptre[sceptre$gRNA_id==guide,]$p_value, 'fdr')
}
  
## add true positives
sceptre$target <- gRNA_ann[match(sceptre$gRNA_id, gRNA_ann$ID),]$target
sceptre$expected_gene <- gRNA_ann[match(sceptre$target, gRNA_ann$target),]$expected_DE_gene
sceptre$class <- gRNA_ann[match(sceptre$target, gRNA_ann$target),]$class
sceptre$sig <- sceptre$FDR < 0.05

# separate NT data
sceptre.nt <- sceptre[sceptre$target == "NT",]
sceptre <- sceptre[sceptre$target != "NT",]

# reorder columns
sceptre <- sceptre[,c(1,2,12:14,3,11,15, 4:10)]
sceptre.nt <- sceptre.nt[,c(1,2,12:14,3,11,15, 4:10)]
```

```{r mast}
files <- list.files(paste0(dir, "results/04_MAST/"))
mast <- list()
for(file in grep("neighbourhood_seed204", files, value=TRUE)){
  mast[[file]] <- readRDS(paste0(dir, "results/04_MAST/", file))
  mast[[file]] <- do.call(rbind, mast[[file]])
}
mast <- as.data.frame(do.call(rbind, mast))

# rename columns
colnames(mast)[1] <- "gene_id"
colnames(mast)[2] <- "p_value"
colnames(mast)[3] <- "logFC"
colnames(mast)[4] <- "logFC.hi"
colnames(mast)[5] <- "logFC.lo"
colnames(mast)[8] <- "gRNA_id"

## add true positives
mast$expected_gene <- gRNA_ann[match(mast$target, gRNA_ann$target),]$expected_DE_gene
mast$class <- gRNA_ann[match(mast$target, gRNA_ann$target),]$class
mast$sig <- mast$FDR <0.05

# separate NT data
mast.nt <- mast[grep("^NT_", mast$target),]
mast <- mast[grep("^NT_", mast$target, invert=TRUE),]

# reorder columns
mast <- mast[,c(1,8,7,9:10,2,6,11, 3:5)]
mast.nt <- mast.nt[,c(1,8,7,9:10,2,6,11, 3:5)]
```

```{r limma}
limma <- readRDS(paste0(dir, "results/04.4_results_limma_neighbourhood.Rds"))
limma <- do.call(rbind, limma)

# rename columns
colnames(limma)[2] <- "gene_id"
colnames(limma)[7] <- "p_value"
colnames(limma)[8] <- "FDR"
colnames(limma)[11] <- "gRNA_id"

## add true positives
limma$expected_gene <- gRNA_ann[match(limma$target, gRNA_ann$target),]$expected_DE_gene
limma$class <- gRNA_ann[match(limma$target, gRNA_ann$target),]$class
limma$sig <- limma$FDR < 0.05

# separate NT data
limma.nt <- limma[grep("^NT_", limma$target),]
limma <- limma[grep("^NT_", limma$target, invert=TRUE),]

# reorder columns
limma <- limma[,c(2,11,10,12:13,7:8,14, 1,4:6,9)]
limma.nt <- limma.nt[,c(2,11,10,12:13,7:8,14, 1,4:6,9)]
```

```{r wilcox}
wilcox <- readRDS(paste0(dir, "results/04.5_results_wilcoxon_neighbourhood.Rds"))
for(i in 1:length(wilcox)){
  wilcox[[i]]$gene_id <- row.names(wilcox[[i]])
}
wilcox <- as.data.frame(do.call(rbind, wilcox))

# rename columns
colnames(wilcox)[2] <- "p_value"
colnames(wilcox)[4] <- "gRNA_id"

## add true positives
wilcox$target <- gRNA_ann[match(wilcox$gRNA, gRNA_ann$ID),]$target
wilcox$expected_gene <- gRNA_ann[match(wilcox$target, gRNA_ann$target),]$expected_DE_gene
wilcox$class <- gRNA_ann[match(wilcox$target, gRNA_ann$target),]$class
wilcox$sig <- wilcox$FDR < 0.05

# separate NT data
wilcox.nt <- wilcox[wilcox$target == "NT",]
wilcox <- wilcox[wilcox$target != "NT",]

# reorder columns
wilcox <- wilcox[,c(5,4,6:8,2:3,9, 1)]
wilcox.nt <- wilcox.nt[,c(5,4,6:8,2:3,9, 1)]
```


#### Detection of true positives {.tabset}

First, we look at whether algorithms are able to identify the expected downregulation of the genes associated with `TSS`, `LCR` and `ENH` perturbations.

##### sceptre

`sceptre` identifies significant differential downregulation for all `LCR` and all but two `TSS` targets, with at least one gRNA. For `ENH` perturbations, 16 (out of 26) also lead to a significant result for at least one gRNA per target.

```{r sceptre_tp, fig.width=5, fig.height=8}
## assess TP recovery
sceptre.exp <- sceptre[!is.na(sceptre$expected_gene),] # separate targets with expected effect
sceptre.tp <- sceptre.exp[sceptre.exp$gene_id == sceptre.exp$expected_gene,]

## visualise as heatmap
pvals <- matrix(-log10(sceptre.tp$FDR), ncol=4, nrow=length(unique(sceptre.tp$target)), byrow = TRUE)
colnames(pvals) <- paste0("gRNA", 1:4)
row.names(pvals) <- unique(sceptre.tp$target)

# annotate by perturbation type
cols <- as.character(GetColors(n=5, scheme = "muted")[c(1,3,5)])
names(cols) <- unique(sceptre.tp$class)
ha  <- rowAnnotation(class = sceptre.tp[seq(1,nrow(sceptre.tp),4),]$class,
                     n_DE = anno_barplot(sapply(seq(1,nrow(sceptre.tp),4), 
                                                   function(x) sum(sceptre.tp[x:(x+3),]$sig & sceptre.tp[x:(x+3),]$z_value < 0))),
                     col = list(class = cols))
Heatmap(pvals,
        col = circlize::colorRamp2(breaks = c(0,-log10(0.05),3,5,10,15),
                                   colors = c("white", brewer.pal(n=9, "BuPu")[3:7])),
        cluster_columns = FALSE, 
        right_annotation = ha,
        split = sceptre.tp[seq(1,nrow(sceptre.tp),4),]$class,
        row_names_gp = gpar(fontsize = 8),
        heatmap_legend_param = list(title = "-log10(FDR)"),
        column_title = "sceptre")
```

The vast majority of perturbation effects are supported by at least 2 independent gRNAs, and over 60% are significant with three or four gRNAs.

```{r sceptre_n_gRNAs_sig}
prop.table(table(n_gRNAs=table(sceptre.tp[sceptre.tp$sig & sceptre.tp$z_value < 0, c('target')])))*100
```

As expected, almost all significant results correspond to gene downregulation, compared to *wild-type* cells. However, a small number of gRNAs induce significant upregulation instead, almost all from `ENH` targets. Generally, a single gRNA shows this effect, but we observe consistent upregulation of *TMSB4X* (with three independent enhancers and 5 different gRNAs) suggesting a possible difference in regulation between K562 and T cells.

```{r sceptre_up}
sceptre.tp[sceptre.tp$sig & sceptre.tp$z_value > 0,]
```

##### MAST

`MAST` also identifies significant downregulation of all `LCR` targets, and all but one `TSS` perturbations. 17 out of 26 `ENH` targets have at least one significant gRNA result.

```{r mast_tp, fig.width=5, fig.height=8}
## assess TP recovery
mast.exp <- mast[!is.na(mast$expected_gene),] # separate targets with expected effect
mast.tp <- mast.exp[mast.exp$gene_id == mast.exp$expected_gene,]

## visualise as heatmap
pvals <- matrix(-log10(mast.tp$FDR), ncol=4, nrow=length(unique(mast.tp$target)), byrow = TRUE)
colnames(pvals) <- paste0("gRNA", 1:4)
row.names(pvals) <- unique(mast.tp$target)
# some p-values are 0; assign smallest non-zero value to avoid Inf
pvals[pvals==Inf] <- -log10(min(mast.tp[mast.tp$FDR > 0,]$FDR))

# annotate by perturbation type
cols <- as.character(GetColors(n=5, scheme = "muted")[c(1,3,5)])
names(cols) <- unique(mast.tp$class)
# and count number of significnat gRNAs
# consider that if all the perturbed cells have 0 count a fold-change is not estimated, but it should be considered as a negative (downregulation) effect
ha  <- rowAnnotation(class = mast.tp[seq(1,nrow(mast.tp),4),]$class,
                     n_DE = anno_barplot(sapply(seq(1,nrow(mast.tp),4), 
                                                   function(x) sum(mast.tp[x:(x+3),]$sig & 
                                                                     (mast.tp[x:(x+3),]$logFC < 0 | is.na(mast.tp[x:(x+3),]$logFC))))),
                     col = list(class = cols))
Heatmap(pvals,
        col = circlize::colorRamp2(breaks = c(0,-log10(0.05),30,50,100,150),
                                   colors = c("white", brewer.pal(n=9, "BuPu")[3:7])),
        cluster_columns = FALSE, 
        right_annotation = ha,
        split = mast.tp[seq(1,nrow(mast.tp),4),]$class,
        row_names_gp = gpar(fontsize = 8),
        heatmap_legend_param = list(title = "-log10(FDR)"),
        column_title = "MAST")
```

Again, the vast majority of perturbation effects are supported by at least 2 independent gRNAs, and in this case over 65% are significant with three or four gRNAs. A larger proportion of targets are supported by higher number of gRNAs compared to `sceptre`, suggesting higher sensitivity.

```{r mast_n_gRNAs_sig}
round(prop.table(table(n_gRNAs=table(mast.tp[mast.tp$sig & mast.tp$logFC < 0, c('target')])))*100, 2)
```

`MAST` also reports a few cases of gene upregulation, and all were also identified by `sceptre`. 

```{r mast_up}
tmp <- mast.tp[mast.tp$sig,]
tmp[which(tmp$logFC > 0),]
```


##### limma

`limma-voom` also identifies significant differential downregulation for all `LCR` and all but two `TSS` targets, with at least one gRNA. For `ENH` perturbations, 20 (out of 26) lead to a significant result for at least one gRNA per target.

```{r limma_tp, fig.width=5, fig.height=8}
## assess TP recovery
limma.exp <- limma[!is.na(limma$expected_gene),] # separate targets with expected effect
limma.tp <- limma.exp[limma.exp$gene_id == limma.exp$expected_gene,]


## visualise as heatmap
pvals <- matrix(-log10(limma.tp$FDR), ncol=4, nrow=length(unique(limma.tp$target)), byrow = TRUE)
colnames(pvals) <- paste0("gRNA", 1:4)
row.names(pvals) <- unique(limma.tp$target)
# some p-values are 0; assign smallest non-zero value to avoid Inf
pvals[pvals==Inf] <- -log10(min(limma.tp[limma.tp$FDR > 0,]$FDR))

# annotate by perturbation type
cols <- as.character(GetColors(n=5, scheme = "muted")[c(1,3,5)])
names(cols) <- unique(limma.tp$class)
ha  <- rowAnnotation(class = limma.tp[seq(1,nrow(limma.tp),4),]$class,
                     n_DE = anno_barplot(sapply(seq(1,nrow(limma.tp),4), 
                                                   function(x) sum(limma.tp[x:(x+3),]$sig & limma.tp[x:(x+3),]$logFC < 0))),
                     col = list(class = cols))
Heatmap(pvals,
        col = circlize::colorRamp2(breaks = c(0,-log10(0.05),30,50,100,150),
                                   colors = c("white", brewer.pal(n=9, "BuPu")[3:7])),
        cluster_columns = FALSE, 
        right_annotation = ha,
        split = limma.tp[seq(1,nrow(limma.tp),4),]$class,
        row_names_gp = gpar(fontsize = 8),
        heatmap_legend_param = list(title = "-log10(FDR)"),
        column_title = "limma-voom")
```

The vast majority of perturbation effects are supported by at least 2 independent gRNAs, and over 60% are significant with three or four gRNAs.

```{r limma_n_gRNAs_sig}
round(prop.table(table(n_gRNAs=table(limma.tp[limma.tp$sig & limma.tp$logFC < 0, c('target')])))*100, 2)
```

Again, almost all significant results correspond to gene downregulation, but a few genes instead show higher expression. All of these *hits* were also identified by `sceptre` and `MAST`. 

```{r limma_up}
limma.tp[limma.tp$sig & limma.tp$logFC > 0,]
```

##### wilcoxon

Results are also very similar when using a `wilcoxon rank sum` test instead. Significant downregulation for at least one gRNA is detected for all `LCR` and all but four `TSS` targets. In this case, 19 out of 26 `ENH` targets have a significant result.

```{r wilcox_tp, fig.width=5, fig.height=8}
## assess TP recovery
wilcox.exp <- wilcox[!is.na(wilcox$expected_gene),] # separate targets with expected effect
wilcox.tp <- wilcox.exp[wilcox.exp$gene_id == wilcox.exp$expected_gene,]

## visualise as heatmap
pvals <- matrix(-log10(wilcox.tp$FDR), ncol=4, nrow=length(unique(wilcox.tp$target)), byrow = TRUE)
colnames(pvals) <- paste0("gRNA", 1:4)
row.names(pvals) <- unique(wilcox.tp$target)

# annotate by perturbation type
cols <- as.character(GetColors(n=5, scheme = "muted")[c(1,3,5)])
names(cols) <- unique(wilcox.tp$class)
ha  <- rowAnnotation(class = wilcox.tp[seq(1,nrow(wilcox.tp),4),]$class,
                     n_DE = anno_barplot(sapply(seq(1,nrow(wilcox.tp),4), 
                                                   function(x) sum(wilcox.tp[x:(x+3),]$sig & wilcox.tp[x:(x+3),]$AUC < 0.5))),
                     col = list(class = cols))
Heatmap(pvals,
        col = circlize::colorRamp2(breaks = c(0,-log10(0.05),30,50,100,150),
                                   colors = c("white", brewer.pal(n=9, "BuPu")[3:7])),
        cluster_columns = FALSE, 
        right_annotation = ha,
        split = wilcox.tp[seq(1,nrow(wilcox.tp),4),]$class,
        row_names_gp = gpar(fontsize = 8),
        heatmap_legend_param = list(title = "-log10(FDR)"),
        column_title = "wilcoxon rank sum test")
```

Most targets are supported by several gRNAs, although the cases with three and four gRNAs is a little lower. 

```{r wilcox_n_gRNAs_sig}
round(prop.table(table(n_gRNAs=table(wilcox.tp[wilcox.tp$sig & wilcox.tp$AUC < 0.5, c('target')])))*100, 2)
```

Again, for the few cases where significant upregulation is detected, most are also reported by the other methods, but there are a few additional *hits*. 

```{r wilcox_up}
wilcox.tp[wilcox.tp$sig & wilcox.tp$AUC > 0.5,]
```

####

---

Overall, all methods are able to recover true positives well, and report largely consistent results. The `wilcoxon rank sum test` recovers slightly fewer significant expected results (`r sum(wilcox.tp$sig)`), and `limma-voom` recovers the most (`r sum(limma.tp$sig)`).

#### Additional DE genes around the target{.tabset}

Next, we check how many other genes are found significantly changing expression in the 2Mb interval around the target.

##### sceptre

- 40% of all gRNAs result in a single DE gene, and it corresponds to the expected gene in most cases (75%).
- Another 40% of all gRNAs result in 2 to 4 DE genes and, again, most of these include the expected gene (79%).
- A small number of gRNAs induce expression changes in a larger (>10) number of genes.

```{r sceptre_total_degs, warning=FALSE, message=FALSE, fig.width=10, fig.height=4}
## total DEGs per gRNA (in positive control perturbations)
n_DE <- as.data.frame(table(sceptre.exp[sceptre.exp$sig,]$gRNA))
colnames(n_DE) <- c("gRNA", "n_DEGs_all")

# exclude TP
tmp <- sceptre.exp[which(sceptre.exp$gene_id != sceptre.exp$expected_gene),]
tmp <- tmp[tmp$sig,]
tmp <- as.data.frame(table(tmp$gRNA_id))
n_DE$other <- tmp[match(as.character(n_DE$gRNA), as.character(tmp$Var1)),2]
n_DE[is.na(n_DE$other),]$other <- 0
n_DE$expected <- n_DE$n_DEGs_all - n_DE$other

df <- reshape2::melt(n_DE[,-2])
df$gRNA <- factor(df$gRNA, levels=unique(df[order(df$value, decreasing = TRUE),]$gRNA))
df$class <- factor(gRNA_ann[match(df$gRNA, gRNA_ann$ID),]$class, levels=c("TSS", "LCR", "ENH"))
colnames(df)[2] <- "DEgene"

ggplot(df, aes(gRNA, value, fill=DEgene)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("grey60", "indianred")) +
  facet_grid(~class, scale="free_x", space="free_x") +
  xlab("all gRNAs with at least one DE gene") +
  ylab("number of DE genes") +
  ggtitle("sceptre") +
  th + theme(axis.text.x = element_blank(),
             axis.ticks.x = element_blank(),
             panel.border = element_rect(fill=NA, size=0.5),
             panel.grid.major.y = element_line(),
             legend.position = "bottom")
```

The gRNAs that result in 10 or more DE genes correspond to `TSS` perturbations and one `ENH` target. Given the `TSS` controls involve proteins important for T cell biology, it is expected that these might have more widespread effects in the transcriptome.

```{r sceptre_many_hits}
many <- n_DE[n_DE$n_DEGs_all >= 10,-3]
many <- cbind(many, gRNA_ann[match(many$gRNA, gRNA_ann$ID),2:4])
many
```

Importantly, the set of *additional* DE genes are most often only identified with one out of the four gRNAs. In contrast, 80% of the expected DE genes are supported by at least 2 independent gRNAs.

```{r sceptre_1Mb_rep}
degs <- table(paste0(sceptre.exp[sceptre.exp$sig,]$gene_id, ".", sceptre.exp[sceptre.exp$sig,]$target))
degs.exp <- table(paste0(sceptre.tp[sceptre.tp$sig,]$gene_id, ".", sceptre.tp[sceptre.tp$sig,]$target))

df <- cbind(1:4,
            round(prop.table(table(degs))*100, 2),
            round(prop.table(table(degs.exp))*100, 2))
colnames(df) <- c("n_sig_gRNAs", "all", "expected")
df
```

This suggests that at least a fraction of the additional DE genes are false positives and/or off-target effects.

##### MAST

Similar results observed with `MAST`.

- 40% of all gRNAs result in a single DE gene, and it corresponds to the expected gene in most cases (72%).
- Another 40% of all gRNAs result in 2 to 4 DE genes and, again, most of these include the expected gene (82%).

However, `MAST` identifies fewer additional *hits* compared to `sceptre`, especially for larger number of DE genes (except for one outlier `ENH` gRNA that returns a very large number of DEGs, the number of gRNAs with 10 or more DEGs is only 7 compared to 15 for `sceptre`).

```{r mast_total_degs, warning=FALSE, message=FALSE, fig.width=10, fig.height=4}
## total DEGs per gRNA (in positive control perturbations)
n_DE <- as.data.frame(table(mast.exp[mast.exp$sig,]$gRNA))
colnames(n_DE) <- c("gRNA", "n_DEGs_all")

# exclude TP
tmp <- mast.exp[which(mast.exp$gene_id != mast.exp$expected_gene),]
tmp <- tmp[tmp$sig,]
tmp <- as.data.frame(table(tmp$gRNA_id))
n_DE$other <- tmp[match(as.character(n_DE$gRNA), as.character(tmp$Var1)),2]
n_DE[is.na(n_DE$other),]$other <- 0
n_DE$expected <- n_DE$n_DEGs_all - n_DE$other

df <- reshape2::melt(n_DE[,-2])
df$gRNA <- factor(df$gRNA, levels=unique(df[order(df$value, decreasing = TRUE),]$gRNA))
df$class <- factor(gRNA_ann[match(df$gRNA, gRNA_ann$ID),]$class, levels=c("TSS", "LCR", "ENH"))
colnames(df)[2] <- "DEgene"

ggplot(df, aes(gRNA, value, fill=DEgene)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("grey60", "indianred")) +
  facet_grid(~class, scale="free_x", space="free_x") +
  xlab("all gRNAs with at least one DE gene") +
  ylab("number of DE genes") +
  ggtitle("MAST") +
  th + theme(axis.text.x = element_blank(),
             axis.ticks.x = element_blank(),
             panel.border = element_rect(fill=NA, size=0.5),
             panel.grid.major.y = element_line(),
             legend.position = "bottom")
```

Again, the gRNAs that result in 10 or more DE genes correspond to `TSS` perturbations and one `ENH` target.

```{r mast_many_hits}
many <- n_DE[n_DE$n_DEGs_all >= 10,-3]
many <- cbind(many, gRNA_ann[match(many$gRNA, gRNA_ann$ID),2:4])
many
```

Again, many of the additional hits are only supported by a single gRNA, whereas expected genes are usually significant with independent gRNAs.

```{r mast_1Mb_rep}
degs <- table(paste0(mast.exp[mast.exp$sig,]$gene_id, ".", mast.exp[mast.exp$sig,]$target))
degs.exp <- table(paste0(mast.tp[mast.tp$sig,]$gene_id, ".", mast.tp[mast.tp$sig,]$target))

df <- cbind(1:4,
            round(prop.table(table(degs))*100, 2),
            round(prop.table(table(degs.exp))*100, 2))
colnames(df) <- c("n_sig_gRNAs", "all", "expected")
df
```

##### limma

Results with `limma-voom` are very similar to those from `MAST`.

- 39% of all gRNAs result in a single DE gene, and it corresponds to the expected gene in most cases (77%).
- Another 43% of all gRNAs result in 2 to 4 DE genes and, again, most of these include the expected gene (82%).
- Similar distribution of number of additional genes as seen with `MAST`.

```{r limma_total_degs, warning=FALSE, message=FALSE, fig.width=10, fig.height=4}
## total DEGs per gRNA (in positive control perturbations)
n_DE <- as.data.frame(table(limma.exp[limma.exp$sig,]$gRNA))
colnames(n_DE) <- c("gRNA", "n_DEGs_all")

# exclude TP
tmp <- limma.exp[which(limma.exp$gene_id != limma.exp$expected_gene),]
tmp <- tmp[tmp$sig,]
tmp <- as.data.frame(table(tmp$gRNA_id))
n_DE$other <- tmp[match(as.character(n_DE$gRNA), as.character(tmp$Var1)),2]
n_DE[is.na(n_DE$other),]$other <- 0
n_DE$expected <- n_DE$n_DEGs_all - n_DE$other

df <- reshape2::melt(n_DE[,-2])
df$gRNA <- factor(df$gRNA, levels=unique(df[order(df$value, decreasing = TRUE),]$gRNA))
df$class <- factor(gRNA_ann[match(df$gRNA, gRNA_ann$ID),]$class, levels=c("TSS", "LCR", "ENH"))
colnames(df)[2] <- "DEgene"

ggplot(df, aes(gRNA, value, fill=DEgene)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("grey60", "indianred")) +
  facet_grid(~class, scale="free_x", space="free_x") +
  xlab("all gRNAs with at least one DE gene") +
  ylab("number of DE genes") +
  ggtitle("limma-voom") +
  th + theme(axis.text.x = element_blank(),
             axis.ticks.x = element_blank(),
             panel.border = element_rect(fill=NA, size=0.5),
             panel.grid.major.y = element_line(),
             legend.position = "bottom")
```

Again, the gRNAs that result in 10 or more DE genes correspond to `TSS` perturbations and one `ENH` target.

```{r limma_many_hits}
many <- n_DE[n_DE$n_DEGs_all >= 10,-3]
many <- cbind(many, gRNA_ann[match(many$gRNA, gRNA_ann$ID),2:4])
many
```

Again, many of the additional hits are only supported by a single gRNA, whereas expected genes are usually significant with independent gRNAs.

```{r limma_1Mb_rep}
degs <- table(paste0(limma.exp[limma.exp$sig,]$gene_id, ".", limma.exp[limma.exp$sig,]$target))
degs.exp <- table(paste0(limma.tp[limma.tp$sig,]$gene_id, ".", limma.tp[limma.tp$sig,]$target))

df <- cbind(1:4,
            round(prop.table(table(degs))*100, 2),
            round(prop.table(table(degs.exp))*100, 2))
colnames(df) <- c("n_sig_gRNAs", "all", "expected")
df
```

##### wilcoxon

Similar results with the `wilcoxon rank sum` test but, again, lower number of additional hits.

- 43% of all gRNAs result in a single DE gene, and it corresponds to the expected gene in most cases (70%).
- Another 42% of all gRNAs result in 2 to 4 DE genes and, again, most of these include the expected gene (81%).

```{r wilcox_total_degs, warning=FALSE, message=FALSE, fig.width=10, fig.height=4}
## total DEGs per gRNA (in positive control perturbations)
n_DE <- as.data.frame(table(wilcox.exp[wilcox.exp$sig,]$gRNA))
colnames(n_DE) <- c("gRNA", "n_DEGs_all")

# exclude TP
tmp <- wilcox.exp[which(wilcox.exp$gene_id != wilcox.exp$expected_gene),]
tmp <- tmp[tmp$sig,]
tmp <- as.data.frame(table(tmp$gRNA_id))
n_DE$other <- tmp[match(as.character(n_DE$gRNA), as.character(tmp$Var1)),2]
n_DE[is.na(n_DE$other),]$other <- 0
n_DE$expected <- n_DE$n_DEGs_all - n_DE$other

df <- reshape2::melt(n_DE[,-2])
df$gRNA <- factor(df$gRNA, levels=unique(df[order(df$value, decreasing = TRUE),]$gRNA))
df$class <- factor(gRNA_ann[match(df$gRNA, gRNA_ann$ID),]$class, levels=c("TSS", "LCR", "ENH"))
colnames(df)[2] <- "DEgene"

ggplot(df, aes(gRNA, value, fill=DEgene)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("grey60", "indianred")) +
  facet_grid(~class, scale="free_x", space="free_x") +
  xlab("all gRNAs with at least one DE gene") +
  ylab("number of DE genes") +
  ggtitle("wilcoxon rank sum") +
  th + theme(axis.text.x = element_blank(),
             axis.ticks.x = element_blank(),
             panel.border = element_rect(fill=NA, size=0.5),
             panel.grid.major.y = element_line(),
             legend.position = "bottom")
```


```{r wilcox_many_hits}
many <- n_DE[n_DE$n_DEGs_all >= 10,-3]
many <- cbind(many, gRNA_ann[match(many$gRNA, gRNA_ann$ID),2:4])
many
```

As mentioned before, the support from 3/4 gRNAs of expected genes is a little lower compared to other methods, and we still observe that many of the additional hits come from a single gRNA.

```{r wilcox_1Mb_rep}
degs <- table(paste0(wilcox.exp[wilcox.exp$sig,]$gene_id, ".", wilcox.exp[wilcox.exp$sig,]$target))
degs.exp <- table(paste0(wilcox.tp[wilcox.tp$sig,]$gene_id, ".", wilcox.tp[wilcox.tp$sig,]$target))

df <- cbind(1:4,
            round(prop.table(table(degs))*100, 2),
            round(prop.table(table(degs.exp))*100, 2))
colnames(df) <- c("n_sig_gRNAs", "all", "expected")
df
```

#### P-value calibration {.tabset}

To get an idea on how likely it is that some of the additional hits are false positives, we check how often do NT gRNAs result in significant results. We test the 35 NT gRNAs against all genes within 1MB of a targeted locus (`r nrow(sceptre.nt)` tests).

##### sceptre

`sceptre` reports `r sum(sceptre.nt$sig)` gene-NTgRNA pairs with a significant p-value, under a FDR of 5%. This indicates `sceptre` does not achieve adequate FDR control. Indeed, the p-values for NT gRNAs are heavily inflated, indicating that `sceptre` is not appropriate for this dataset, where the MOI is 1.

```{r sceptre_nt, fig.width=5, fig.height=4}
## code adapted from https://github.com/Katsevich-Lab/sceptre-manuscript/blob/master/sceptre_paper/
## check p-value callibration
ci <- 0.95
p_thresh <- 1e-8

pvals_nt <- sceptre.nt[,c('gene_id', 'gRNA_id', 'p_value')]

## compute expected p-values for QQplots
df <- pvals_nt %>%
    mutate(r = rank(p_value), expected = ppoints(n())[r],
           clower = qbeta(p=(1-ci)/2, shape1 = r, shape2 = n()+1-r),
           cupper = qbeta(p=(1+ci)/2, shape1 = r, shape2 = n()+1-r)) %>%
  ungroup() %>% mutate()

df %>% 
  mutate(p_value = ifelse(p_value < p_thresh, p_thresh, p_value)) %>%
  ggplot(aes(x = expected, y = p_value, ymin = clower, ymax = cupper)) +
    geom_point(size = 1, alpha = 0.5) +
    geom_ribbon(alpha = 0.2) +
    geom_abline(intercept = 0, slope = 1) +
    scale_x_continuous(trans = revlog_trans(base = 10)) + 
    scale_y_continuous(trans = revlog_trans(base = 10)) +
    xlab(expression(paste("Expected null p-value"))) +
    ylab(expression(paste("Observed p-value"))) +
    ggtitle("sceptre") +
    th
```

##### MAST

`MAST` reports `r sum(mast.nt$sig)` gene-NTgRNA pairs with a significant p-value, under a FDR of 5%. A bit better than `sceptre` but still showing lack of adequate FDR control, and heavily inflated p-values.

```{r mast_nt, fig.width=5, fig.height=4}
## check p-value callibration
ci <- 0.95
p_thresh <- 1e-8

pvals_nt <- mast.nt[,c('gene_id', 'gRNA_id', 'p_value')]

## compute expected p-values for QQplots
df <- pvals_nt %>%
    mutate(r = rank(p_value), expected = ppoints(n())[r],
           clower = qbeta(p=(1-ci)/2, shape1 = r, shape2 = n()+1-r),
           cupper = qbeta(p=(1+ci)/2, shape1 = r, shape2 = n()+1-r)) %>%
  ungroup() %>% mutate()

df %>% 
  mutate(p_value = ifelse(p_value < p_thresh, p_thresh, p_value)) %>%
  ggplot(aes(x = expected, y = p_value, ymin = clower, ymax = cupper)) +
    geom_point(size = 1, alpha = 0.5) +
    geom_ribbon(alpha = 0.2) +
    geom_abline(intercept = 0, slope = 1) +
    scale_x_continuous(trans = revlog_trans(base = 10)) + 
    scale_y_continuous(trans = revlog_trans(base = 10)) +
    xlab(expression(paste("Expected null p-value"))) +
    ylab(expression(paste("Observed p-value"))) +
    ggtitle("MAST") +
    th
```

##### limma

`limma-voom` reports `r sum(limma.nt$sig)` gene-NTgRNA pairs with a significant p-value, and also shows heavily inflated p-values.

```{r limma_nt, fig.width=5, fig.height=4}
## check p-value callibration
ci <- 0.95
p_thresh <- 1e-8

pvals_nt <- limma.nt[,c('gene_id', 'gRNA_id', 'p_value')]

## compute expected p-values for QQplots
df <- pvals_nt %>%
    mutate(r = rank(p_value), expected = ppoints(n())[r],
           clower = qbeta(p=(1-ci)/2, shape1 = r, shape2 = n()+1-r),
           cupper = qbeta(p=(1+ci)/2, shape1 = r, shape2 = n()+1-r)) %>%
  ungroup() %>% mutate()

df %>% 
  mutate(p_value = ifelse(p_value < p_thresh, p_thresh, p_value)) %>%
  ggplot(aes(x = expected, y = p_value, ymin = clower, ymax = cupper)) +
    geom_point(size = 1, alpha = 0.5) +
    geom_ribbon(alpha = 0.2) +
    geom_abline(intercept = 0, slope = 1) +
    scale_x_continuous(trans = revlog_trans(base = 10)) + 
    scale_y_continuous(trans = revlog_trans(base = 10)) +
    xlab(expression(paste("Expected null p-value"))) +
    ylab(expression(paste("Observed p-value"))) +
    ggtitle("limma-voom") +
    th
```

##### wilcoxon

The `wilcoxon rank sum` test reports `r sum(wilcox.nt$sig)` gene-NTgRNA pairs with a significant p-value, the second lowest total. 

```{r wilcox_nt, fig.width=5, fig.height=4}
## check p-value callibration
ci <- 0.95
p_thresh <- 1e-8

pvals_nt <- wilcox.nt[,c('gene_id', 'gRNA_id', 'p_value')]

## compute expected p-values for QQplots
df <- pvals_nt %>%
    mutate(r = rank(p_value), expected = ppoints(n())[r],
           clower = qbeta(p=(1-ci)/2, shape1 = r, shape2 = n()+1-r),
           cupper = qbeta(p=(1+ci)/2, shape1 = r, shape2 = n()+1-r)) %>%
  ungroup() %>% mutate()

df %>% 
  mutate(p_value = ifelse(p_value < p_thresh, p_thresh, p_value)) %>%
  ggplot(aes(x = expected, y = p_value, ymin = clower, ymax = cupper)) +
    geom_point(size = 1, alpha = 0.5) +
    geom_ribbon(alpha = 0.2) +
    geom_abline(intercept = 0, slope = 1) +
    scale_x_continuous(trans = revlog_trans(base = 10)) + 
    scale_y_continuous(trans = revlog_trans(base = 10)) +
    xlab(expression(paste("Expected null p-value"))) +
    ylab(expression(paste("Observed p-value"))) +
    ggtitle("wilcoxon rank sum") +
    th
```


### Method comparison

#### True positive recovery

```{r compare_methods_TP, fig.width=7, fig.height=3}
compare <- gRNA_ann[,c(1,4,3,5)]
compare$sceptre <- sceptre[match(paste0(compare$ID, ".", compare$expected_DE_gene),
                                 paste0(sceptre$gRNA_id, ".", sceptre$gene_id)),]$FDR
compare$sceptre_sig <- ifelse(compare$sceptre < 0.05, 1, 0)

compare$mast <- mast[match(paste0(compare$ID, ".", compare$expected_DE_gene),
                           paste0(mast$gRNA_id, ".", mast$gene_id)),]$FDR
compare$mast_sig <- ifelse(compare$mast < 0.05, 1, 0)

compare$limma <- limma[match(paste0(compare$ID, ".", compare$expected_DE_gene),
                             paste0(limma$gRNA_id, ".", limma$gene_id)),]$FDR
compare$limma_sig <- ifelse(compare$limma < 0.05, 1, 0)

compare$wilcoxon <- wilcox[match(paste0(compare$ID, ".", compare$expected_DE_gene),
                                 paste0(wilcox$gRNA_id, ".", wilcox$gene_id)),]$FDR
compare$wilcoxon_sig <- ifelse(compare$wilcoxon < 0.05, 1, 0)

compare <- compare[!is.na(compare$sceptre),]
compare$sum <- rowSums(compare[,seq(6,12,2)])
```

All methods are able to recover true positives and report similar results. From **`r nrow(compare)` different gRNA-expected_gene pairs**:

- `r sum(compare$sum != 0)` are deemed significant by at least one method.
  + From the `r sum(compare$sum == 0)` pairs without any significant results, the majority correspond to `ENH` perturbations, but there are also a number of `TSS` pairs.
  
```{r not_sig}
table(compare[compare$sum==0,]$class)
```
  
- 
  + The `ENH` gRNAs without detected effects correspond to `r length(unique(compare[compare$sum==0 & compare$class == "ENH",]$target))` different targets. From these, two thirds have significant results for other gRNAs targeting the same locus, suggesting that the lack of results stem from non-functional gRNAs and/or small effects that are harder to detect. Instead, 6 targets have at most a single significant gRNA and might correspond to non-functional enhancers in T cells, or very small effects.
  
```{r not_sig_enh}
tmp <- table(n_gRNAs_without_DEGs=compare[compare$sum==0 & compare$class == "ENH",]$target)
tmp[order(tmp)]
```

- 
  + The `TSS` gRNAs without detected DE results correspond to `r length(unique(compare[compare$sum==0 & compare$class == "TSS",]$target))` different targets. In this case, all but one of the targets have significant results for other gRNAs, suggesting the lack of hits derive from non-functional gRNAs.

```{r not_sig_tss}
tmp <- table(n_gRNAs_without_DEGs=compare[compare$sum==0 & compare$class == "TSS",]$target)
tmp[order(tmp)]
```

- Each method reports ~150 significant results.
- 127 gRNA-gene pairs are detected by all four methods.
  + A further 14 are identified by all except the `wilcoxon rank sum test`. 
- `sceptre` misses 18 hits recovered by other methods.
- All methods except `MAST` show a small number of hits that are not reported by any other method.

```{r compare_methods_upset, fig.width=7, fig.height=3}
upset(compare[, c(seq(6,12,2),3)], order.by = "freq",
      queries = list(
        list(query = elements, 
             params = list("class", c("TSS","ENH","LCR")), query.name = "LCR",
             color = GetColors(n=5, scheme = "muted")[5], active = TRUE),
        list(query = elements, 
             params = list("class", c("TSS","ENH")), query.name = "ENH",
             color = GetColors(n=5, scheme = "muted")[1], active = TRUE),
        list(query = elements, 
             params = list("class", "TSS"), query.name = "TSS",
             color = GetColors(n=5, scheme = "muted")[3], active = TRUE)),
      query.legend="bottom")
```

When looking at the groups of genes that are detected by only a subset of methods, compared to those detected by all (or none), we observe that:

- As expected, pairs that are not detected by any method tend to involve genes expressed at lower levels and gRNAs detected in lower number of cells.
- Pairs missed by the `wilcoxon rank sum` test involve genes with low expression and variability.
- Pairs that are not reported by `sceptre` involve genes with higher technical variability.

- `MAST` and `limma-voom` are able to report significant findings for genes expressed at very low levels, also observed for pairs reported by `limma-voom` only.
- Pairs reported only by `limma-voom` involve gRNAs detected in smaller numbers of cells. It is likely that with better gRNA recovery the other methods can pick them up too.
- Pairs reported only by `sceptre` involve genes expressed at higher levels, but with lower variability.
- Pairs reported only by the `wilcoxon rank sum` test also show high variability.

```{r compare_subsets, message=FALSE, warning=FALSE, fig.width=10, fig.height=7}
## annotate the hits specific to some methods
compare$group <- NA
compare[compare$sum==0,]$group <- "none"
compare[compare$sum==4,]$group <- "all"
compare[compare$sum==1 & compare$sceptre_sig,]$group <- "sceptre_only"
compare[compare$sum==1 & compare$limma_sig,]$group <- "limma_only"
compare[compare$sum==1 & compare$wilcoxon_sig,]$group <- "wilcox_only"
compare[compare$sum==3 & !compare$sceptre_sig,]$group <- "sceptre_miss"
compare[compare$sum==3 & !compare$wilcoxon_sig,]$group <- "wilcox_miss"
compare[compare$sum==2 & compare$mast_sig & compare$limma_sig,]$group <- "mast_limma"

## correlate to features that could explain (lack of) detection
var_decomp <- modelGeneVar(sce, BPPARAM = bpp)

df <- compare[,c(1:4,ncol(compare))]
df <- cbind(df, var_decomp[match(df$expected_DE_gene, row.names(var_decomp)), 1:4])
df <- df[!is.na(df$group),] # for simplicity, remove the groups that have very low counts

df$group <- factor(df$group, levels=c("none", "all", "wilcox_miss", "sceptre_miss", "mast_limma", 
                                      paste0(c("limma", "sceptre", "wilcox"), "_only")))
df$n_cells <- ncells_per_gRNA[df$ID]

plots <- list()
plots[[1]] <- ggplot(df, aes(group, n_cells)) +
  scale_y_log10() +
  geom_violin(scale="width") +
  geom_boxplot(width=0.1) +
  facet_wrap(~group, scale="free_x", ncol=8) +
  xlab("") +
  ylab("cells per gRNA") +
  th + theme(axis.text.x = element_blank(),
             axis.ticks.x = element_blank(),
             panel.grid.major.y = element_line())
plots[[2]] <- ggplot(df, aes(group, mean)) +
  geom_violin(scale="width") +
  geom_boxplot(width=0.1) +
  facet_wrap(~group, scale="free_x", ncol=8) +
  xlab("") +
  ylab(expression('log'[2]*' mean expression')) +
  th + theme(axis.text.x = element_blank(),
             axis.ticks.x = element_blank(),
             strip.background = element_blank(),
             strip.text.x = element_blank(),
             panel.grid.major.y = element_line())
tmp <- reshape2::melt(df[,c(5,7:9)])
plots[[3]] <- ggplot(tmp, aes(group, value, colour=variable)) +
  geom_violin(scale="width") +
  geom_boxplot(width=0.1, position = position_dodge(0.9)) +
  scale_color_manual(values = c(total="grey30", tech="grey80", bio="indianred2")) +
  facet_wrap(~group, scale="free_x", ncol=8) +
  xlab("") +
  ylab("variation") +
  labs(colour="") +
  th + theme(axis.text.x = element_blank(),
             axis.ticks.x = element_blank(),
             panel.grid.major.y = element_line(),
             strip.background = element_blank(),
             strip.text.x = element_blank(), 
             legend.position = "bottom")
ggarrange(plotlist = plots, ncol=1, align="v", heights = c(0.35,0.3,0.4))
```

#### Other DE genes

```{r compare_methods_other, fig.width=7, fig.height=3}
compare <- data.frame(pair = unique(c(paste(sceptre.exp[sceptre.exp$sig,]$gRNA_id, sceptre.exp[sceptre.exp$sig,]$gene_id, sep="|"),
                                      paste(mast.exp[mast.exp$sig,]$gRNA_id, mast.exp[mast.exp$sig,]$gene_id, sep="|"),
                                      paste(limma.exp[limma.exp$sig,]$gRNA_id, limma.exp[limma.exp$sig,]$gene_id, sep="|"),
                                      paste(wilcox.exp[wilcox.exp$sig,]$gRNA_id, wilcox.exp[wilcox.exp$sig,]$gene_id, sep="|"))))
compare$gRNA_id <- unlist(lapply(strsplit(compare$pair, "|", fixed=TRUE), '[[', 1))
compare$gene_id <- unlist(lapply(strsplit(compare$pair, "|", fixed=TRUE), '[[', 2))
compare$target <- gRNA_ann[match(compare$gRNA_id, gRNA_ann$ID),]$target
compare$class <- gRNA_ann[match(compare$gRNA_id, gRNA_ann$ID),]$class
compare$expected_DE_gene <- gRNA_ann[match(compare$gRNA_id, gRNA_ann$ID),]$expected_DE_gene
compare$expected <- compare$gene_id == compare$expected_DE_gene

compare$sceptre <- sceptre[match(compare$pair, paste0(sceptre$gRNA_id, "|", sceptre$gene_id)),]$FDR
compare$sceptre_sig <- ifelse(compare$sceptre < 0.05, 1, 0)

compare$mast <- mast[match(compare$pair, paste0(mast$gRNA_id, "|", mast$gene_id)),]$FDR
compare$mast_sig <- ifelse(compare$mast < 0.05, 1, 0)

compare$limma <- limma[match(compare$pair, paste0(limma$gRNA_id, "|", limma$gene_id)),]$FDR
compare$limma_sig <- ifelse(compare$limma < 0.05, 1, 0)

compare$wilcoxon <- wilcox[match(compare$pair, paste0(wilcox$gRNA_id, "|", wilcox$gene_id)),]$FDR
compare$wilcoxon_sig <- ifelse(compare$wilcoxon < 0.05, 1, 0)

compare$sum <- rowSums(compare[,seq(9,15,2)])
```

Overall, there are `r nrow(compare)` gRNA-gene pairs that are deemed significant by at least one method. From these, `r sum(compare$expected)` involve the expected gene, leaving a large number of *additional* DE genes in the vicinity of the targets. This set of additional DEGs contains a subset of 170 pairs that are reported by all methods. However, in contrast to the expected effects, much higher numbers of pairs are reported by a single method, indicative of likely false positives. `sceptre` and the `wilcoxon rank sum` test report substantially higher number of these *private* hits, compared to `limma-voom` and `MAST`.

```{r compare_methods_other_upset, fig.width=7, fig.height=3}
upset(compare[!compare$expected, c(seq(9,15,2),5)], order.by = "freq",
      queries = list(
        list(query = elements, 
             params = list("class", c("TSS","ENH","LCR")), query.name = "LCR",
             color = GetColors(n=5, scheme = "muted")[5], active = TRUE),
        list(query = elements, 
             params = list("class", c("TSS","ENH")), query.name = "ENH",
             color = GetColors(n=5, scheme = "muted")[1], active = TRUE),
        list(query = elements, 
             params = list("class", "TSS"), query.name = "TSS",
             color = GetColors(n=5, scheme = "muted")[3], active = TRUE)),
      query.legend="bottom")
```

Consistently we can see that

- DE genes called by several methods are mostly supported by at least 2 independent gRNAs when grouped by target.
- In contrast, DE genes that are identified by a single method have a large fraction of hits identified with a single gRNA, thus more likely to be false positive or off-target effects. 
- DE genes reported by only two methods still show a large fraction of cases supported by independent gRNAs, but have lower proportion of 3 and 4-gRNA support, indicating these might include smaller effects that are harder to identify.

```{r compare_methods_other_ngrnas, fig.width=7, fig.height=3}
## add number of gRNAs supporting each hit
# group by target
compare$parir_tgt <- paste(compare$target, compare$gene_id, sep='|')

# use the maximum achieved by any method
tmp <- rbind(sceptre = as.data.frame(table(compare[compare$sceptre_sig==1,]$parir_tgt)),
            mast = as.data.frame(table(compare[compare$mast_sig==1,]$parir_tgt)),
            limma = as.data.frame(table(compare[compare$limma_sig==1,]$parir_tgt)),
            wilcox = as.data.frame(table(compare[compare$wilcoxon_sig==1,]$parir_tgt)))

tmp <- tmp %>% group_by(Var1) %>% summarise(max = max(Freq))
compare$n_gRNAs <- tmp[match(compare$parir_tgt, tmp$Var1),]$max

# replot
upset(compare[!compare$expected, c(seq(9,15,2),18)], order.by = "freq",
      queries = list(
        list(query = elements, 
             params = list("n_gRNAs", c('1','2','3','4')), query.name = '4',
             color = GetColors(n=11, scheme = "sunset")[11], active = TRUE),
        list(query = elements, 
             params = list("n_gRNAs", c('1','2','3')), query.name = '3',
             color = GetColors(n=11, scheme = "sunset")[10], active = TRUE),
        list(query = elements, 
             params = list("n_gRNAs", c('1','2')), query.name = '2',
             color = GetColors(n=11, scheme = "sunset")[8], active = TRUE),
        list(query = elements, 
             params = list("n_gRNAs", c('1')), query.name = '1',
             color = GetColors(n=11, scheme = "sunset")[6], active = TRUE)),
      query.legend="bottom")
```

### Conclusions

Overall, we can conclude that

- all methods are good at detecting true positives, and do so consistently.
- all methods also report many additional hits. A fraction of these are likely to be false positives since p-values are miscalibrated.


```{r save}
write.table(sceptre, paste0(dir, "results/05_DE_results.sceptre.tsv"), quote = FALSE, sep="\t", row.names = FALSE)
write.table(mast, paste0(dir, "results/05_DE_results.mast.tsv"), quote = FALSE, sep="\t", row.names = FALSE)
write.table(limma, paste0(dir, "results/05_DE_results.limma.tsv"), quote = FALSE, sep="\t", row.names = FALSE)
write.table(wilcox, paste0(dir, "results/05_DE_results.wilcox.tsv"), quote = FALSE, sep="\t", row.names = FALSE)

write.table(sceptre.nt, paste0(dir, "results/05_DE_results_NT.sceptre.tsv"), quote = FALSE, sep="\t", row.names = FALSE)
write.table(mast.nt, paste0(dir, "results/05_DE_results_NT.mast.tsv"), quote = FALSE, sep="\t", row.names = FALSE)
write.table(limma.nt, paste0(dir, "results/05_DE_results_NT.limma.tsv"), quote = FALSE, sep="\t", row.names = FALSE)
write.table(wilcox.nt, paste0(dir, "results/05_DE_results_NT.wilcox.tsv"), quote = FALSE, sep="\t", row.names = FALSE)

write.table(compare, paste0(dir, "results/05_DEGs_compareMethods.tsv"), quote = FALSE, sep="\t", row.names = FALSE)
```


```{r info}
sessionInfo()
```

